docs/writing_plugins.md

```diff
@@ -0,0 +1,550 @@
+4.  Specify this in your `package.json` file.
+
+**NOTE**:  Plugin authors may optionally implement a JavaScript generator function, a declaration file generator function, or a transpile function for more 
+fine-grained control over the generation process.  See [API](#api) for more information.
+
+The following instructions will walk authors through the process outlined above.
+
+### Generating Code
+
+Generator functions are functions that are used to generate the actual file content parsed from protobuf files.  There are three that can be implemented, corresponding to the three possible target outputs for plugins:
+
+| Target Out | Function |
+| :--- | :--- |
+| `ts` | [`generateTs`](#generatets) |
+| `js` | [`generateJs`](#generatejs) |
+| `dts` | [`generateDts`](#generatedts) |
+
+Of the three, only [`generateTs`](#generatets) is required.  These functions will be passed as part of your plugin initialization and as the plugin runs, the framework will invoke the functions depending on which target outputs were specified by the plugin consumer.  
+
+Since [`generateJs`](#generatejs) and [`generateDts`](#generatedts) are both optional, if they are not provided, the plugin framework will attempt to transpile your generated TypeScript files to generate any desired `js` or `dts` outputs if necessary.  See the [`PluginInit`](#plugininit) object definition for more information.
+
+The generator functions are invoked by the plugin framework with a parameter of type [`Schema`](#schema).  This object contains the information needed to generate code.  In addition to the [`CodeGeneratorRequest`](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/compiler/plugin.proto) that is standard when working with protoc plugins, the [`Schema`](#schema) object also contains some convenience interfaces that make it a bit easier to work with the various descriptor objects.  
+
+For example, the [`Schema`](#schema) object contains a [`files`](#files) property, which is a list of `DescFile` objects representing the files requested to be generated.  The first thing you will most likely do in your generator function is iterate over this list and issue a call to the [`generateFile`](#generatefile) function for each file object.  This will return a generated file object containing a `print` function which you can then use to "print" to the file.
+
+Each `DescFile` object in the [`files`](#files) property contains all enums and messages needed to begin generating code:
+
+```ts
+// protoc-gen-foo-plugin.js
+
+function generateTs(schema: Schema) {
+   for (const file of schema.files) {
+     const f = schema.generateFile(file.name + "_pb.ts");
+  
+     for (const enumeration of file.enums) {
+        f.print(`// generating enums from ${file.name}`);
+        ...
+     }
+  
+     for (const message of file.messages) {
+        f.print(`// generating messages from ${file.name}`);
+        ...
+     }
```

Comment:
Messages can contain other message definitions, and enum definitions. We should call out that this example does not generate all elements.

This is actually something we could support better, by providing a complete list of all messages an a list of all enumerations in a file 🤔